fix bugs in condition mapper

hhkwong opened the pull request.

hhkwong commented on src/main/java/com/salesforce/dynamodbv2/mt/mappers/sharedtable/impl/ConditionMapper.java:
```diff
@@ -16,14 +16,21 @@
+     * to the corresponding physical update request.
      */
-    void apply(RequestWrapper request);
+    void applyForUpdate(RequestWrapper request);
```
Handling update requests is the most difficult and has more complications: each assignment in the update expression can potentially fan out, and the update expression and filter expression may have overlapping field references. (Query requests also have possibly 2 expressions, but each can be translated independently, because the field references can't overlap -- I think.)

busjaeger replied:
By `filter expression` for updates do you mean `ConditionExpression`?

hhkwong replied:
Yes that's what I meant

hhkwong commented on src/main/java/com/salesforce/dynamodbv2/mt/mappers/sharedtable/impl/ConditionMapper.java:
```diff
@@ -16,14 +16,21 @@
      * based on the given virtual secondary index if it's provided, or if not, on the virtual table primary key.
      */
-    void apply(RequestWrapper request, @Nullable DynamoSecondaryIndex virtualSecondaryIndex);
+    void applyToKeyCondition(RequestWrapper request, @Nullable DynamoSecondaryIndex virtualSecondaryIndex);
```
Used to map key condition expression in query requests only

hhkwong commented on src/main/java/com/salesforce/dynamodbv2/mt/mappers/sharedtable/impl/ConditionMapper.java:
```diff
@@ -16,14 +16,21 @@
+     * Maps the filter expression in the given virtual request to the corresponding physical filter expression,
+     * applying the change to either the primary or secondary expression in the request wrapper.
+     */
+    void applyToFilterExpression(RequestWrapper request, boolean isPrimaryExpression);
```
Used to map filter/condition expressions in a put/delete/scan/query request, which in terms of RequestWrapper is primary in the case of put/delete/scan, and secondary in a query request.

busjaeger replied:
Seems like it may be useful to revisit the `RequestWrapper` abstraction in the context of expression mapping to make this interface more explicit (in a future change).

hhkwong replied:
Yeah... I was considering that too but ended up not doing it and limiting the amount of changes in this PR

hhkwong commented on src/main/java/com/salesforce/dynamodbv2/mt/mappers/sharedtable/impl/RandomPartitioningConditionMapper.java:
```diff
@@ -30,143 +38,272 @@
+        if (conditionExpression != null) {
+            conditionExpression = convertFieldNameLiteralsToExpressionNames(conditionExpression, request,
+                virtualFieldNames);
+            conditionExpression = makePlaceholdersDistinct(updateExpression, conditionExpression, request);
```
To make our lives easier, if the condition expression has overlapping attribute name or value references with the update expression, create more placeholder references so there is no overlap

busjaeger replied:
Why is this helpful? I.e., what problem does it create to have overlapping attribute name/value placeholders?

hhkwong replied:
Say that in the virtual update request, the update expression is `SET #fieldA = :value1`, and the condition expression is `#fieldB = :value1`, with attribute name map `{fieldA -> A, fieldB -> B}` and attribute value map `{value1 -> x}`. The virtual table has 2 GSIs defined, GSI1 has HK=A, and GSI2 has HK=C and RK=B.

Mapping the update expression, we'd apply the GSI1 FieldMapping for field A, giving us `SET #fieldA = :value1`, with name map `{fieldA -> gsi1HK}`, and value map `{value1 -> ctx/table/x}`.

Mapping the condition expression, if we did this independently without considering the update expression, we'd apply the GSI2 mapping for B, giving us `#fieldB = :value1`, with name map `{#fieldB -> gsi2RK}`, and value map `{value1 -> x}`.

But then we have conflicting value mappings for the `value1` placeholder and we'd have to deal with this somehow. I think we could have a similar scenario where the same field placeholder is used in the two expressions. Making the two expressions not share any field or value placeholders in the first place would make this easy and we can transform the two expressions independently.

busjaeger replied:
Thanks that helps a lot!

hhkwong commented on src/main/java/com/salesforce/dynamodbv2/mt/mappers/sharedtable/impl/RandomPartitioningConditionMapper.java:
```diff
@@ -30,143 +38,272 @@
+        String expression = isPrimaryExpression ? request.getPrimaryExpression() : request.getFilterExpression();
+
+        // for conditional / filter expressions, for each indexed field, we only need to use one of its field mappings,
+        // since we're supporting only equals expressions
```
This comment is outdated, before I added support for non-equals comparison operators when using non-legacy filter expressions... I'm not sure there's a bug. I'm confused because I'd expect there to be a bug if we do >, >=, <, <= comparisons on a number RK, when using binary encoding, because I thought we didn't have a binary encoding for numbers that preserves order yet...

hhkwong replied:
On okay I was confused because I was thinking currently mt-dynamo already supports mapping a virtual range key of type N to a physical range key of type B. But that's not true and the range key types have to be exactly the same right now. As long as we only allow RK mappings where the order is preserved, we can use any of the index FieldMappings to map comparison clauses for a given virtual field.

hhkwong commented on src/main/java/com/salesforce/dynamodbv2/mt/mappers/sharedtable/impl/RandomPartitioningConditionMapper.java:
```diff
@@ -30,143 +38,272 @@
+                                       FieldMapping fieldMapping) {
+        if (expression != null) {
+            String virtualField = fieldMapping.getSource().getName();
+            Set<String> fieldPlaceholders = getPlaceholdersForVirtualField(virtualField, request); // {#field1}
```
One of the bugs was that if there are multiple placeholders for the same field (in the same expression or across two expressions) in a request, we would only look at the first placeholder we find, and map its attribute name and possibly comparison/assignment value. So now we look at all placeholders for this virtual field.

hhkwong commented on src/main/java/com/salesforce/dynamodbv2/mt/mappers/sharedtable/impl/RandomPartitioningConditionMapper.java:
```diff
@@ -30,143 +38,272 @@
+            String virtualField = fieldMapping.getSource().getName();
+            Set<String> fieldPlaceholders = getPlaceholdersForVirtualField(virtualField, request); // {#field1}
+
+            for (String fieldPlaceholder : fieldPlaceholders) {
```
The diff here is weird, but given a placeholder and a field mapping, the logic is basically the same as before

hhkwong commented on src/main/java/com/salesforce/dynamodbv2/mt/mappers/sharedtable/impl/RandomPartitioningConditionMapper.java:
```diff
@@ -30,143 +38,272 @@
+            request.setFilterExpression(expression);
         }
+    }
+
```
The method below is used by applyToFilterExpression() and by applyToKeyCondition() -- will add comment in code

hhkwong commented on src/main/java/com/salesforce/dynamodbv2/mt/mappers/sharedtable/impl/RandomPartitioningConditionMapper.java:
```diff
@@ -30,143 +38,272 @@
-                            ? fieldMapper.apply(fieldMapping, virtualAttr) // {S: ctx.virtualTable.hkValue,}
-                            : virtualAttr;
-                    request.putExpressionAttributeValue(virtualValuePlaceholder, physicalAttr);
+
```
Below is how we fan out from one field assignment in the virtual SET expression to possibly multiple assignments in the physical SET expression

hhkwong commented on src/main/java/com/salesforce/dynamodbv2/mt/mappers/sharedtable/impl/RandomPartitioningConditionMapper.java:
```diff
@@ -30,143 +38,272 @@
-            AtomicInteger counter = new AtomicInteger(1);
-            for (FieldMapping fieldMapping : fieldMappings) {
-                String virtualFieldName = fieldMapping.getSource().getName(); // "field"
-                String toFind = " " + virtualFieldName + " ="; // " field ="
```
This wasn't working if the virtual field literal is at the beginning of the expression without a space before it

hhkwong commented on src/test/java/com/salesforce/dynamodbv2/UpdateTest.java:
```diff
@@ -69,6 +79,30 @@ void update(TestArgument testArgument) {
 
+    @ParameterizedTest
+    @ArgumentsSource(DefaultArgumentProvider.class)
+    void updatePrimaryKeyFieldFails(TestArgument testArgument) {
```
New test case: can't refer to table PK fields in an update expression and throw exception consistent with single-tenant dynamo (though I got the error message from local dynamo and I'm not sure if that's necessarily the same as what AWS dynamo would give?)

hhkwong commented on src/test/java/com/salesforce/dynamodbv2/UpdateTest.java:
```diff
@@ -122,56 +156,77 @@ void updateConditionalOnHkSuccess(TestArgument testArgument) {
     @ParameterizedTest(name = "{arguments}")
     @ArgumentsSource(DefaultArgumentProvider.class)
-    void updateConditionalOnGsiHkSuccess(TestArgument testArgument) {
+    void updateConditionalOnHkSuccessWithLiterals(TestArgument testArgument) {
```
New test case: same as existing test case that does conditional update, but using literals instead of placeholders

hhkwong replied:
The diff is weird because I swapped this test case with updateConditionalOnGsiHkSuccess() so the former is adjacent to the related test case updateConditionalOnHkSuccess()...

hhkwong commented on src/test/java/com/salesforce/dynamodbv2/UpdateTest.java:
```diff
@@ -122,56 +156,77 @@ void updateConditionalOnHkSuccess(TestArgument testArgument) {
     @ParameterizedTest(name = "{arguments}")
     @ArgumentsSource(DefaultArgumentProvider.class)
-    void updateConditionalOnHkWithLiteralsSuccess(TestArgument testArgument) {
+    void updateConditionalOnGsiRkSuccess_usePlaceholder(TestArgument testArgument) {
```
Expanded existing test case updateConditionalOnGsiRkSuccess into 2 flavors, one using a placeholder (same as before) and one using the literal

hhkwong commented on src/test/java/com/salesforce/dynamodbv2/UpdateTest.java:
```diff
@@ -208,74 +263,164 @@ void updateConditionalFail(TestArgument testArgument) {
     @ArgumentsSource(DefaultArgumentProvider.class)
-    void updateHkRkTableWithGsi(TestArgument testArgument) {
-        runUpdateHkRkTableWithGsiTest(testArgument, TABLE3);
+    void updateGsiFieldAndQueryGsi(TestArgument testArgument) {
```
The bug that broke ZOS and what we originally talked about...

hhkwong commented on src/test/java/com/salesforce/dynamodbv2/UpdateTest.java:
```diff
@@ -208,74 +263,164 @@ void updateConditionalFail(TestArgument testArgument) {
-    }
-
-    private void runUpdateHkRkTableWithGsiTest(TestArgument testArgument, String tableName) {
+    void updateFieldInMultipleGsis(TestArgument testArgument) {
```
New test case where we update a field that is used in multiple GSIs and requires us to fan out. At the end we query both indexes to make sure both queries work.

hhkwong commented on src/test/java/com/salesforce/dynamodbv2/testsupport/DefaultTestSetup.java:
```diff
@@ -185,7 +186,7 @@ protected MtAmazonDynamoDbContextProvider getMtContext() {
                 .withTableName(TABLE1)
                 .withAttributeDefinitions(new AttributeDefinition(HASH_KEY_FIELD, hashKeyAttrType))
-                .withKeySchema(new KeySchemaElement(HASH_KEY_FIELD, KeyType.HASH))
+                .withKeySchema(getKeySchema(HASH_KEY_FIELD, HASH))
```
Most changes in this file are just refactoring and don't change logic. The only logical change is that in TABLE5, I'm adding 2 new GSIs that share a common non-PK field, so I can write the test on updating a field used in multiple GSIs.

msgroi commented on src/main/java/com/salesforce/dynamodbv2/mt/mappers/sharedtable/impl/ConditionMapper.java:
```diff
@@ -16,14 +16,21 @@
 interface ConditionMapper {
```
I feel like some tests on these methods directly might help understand the code better.  Perhaps it's not necessary now and would be a further distraction.  I realize I should have given myself this advice when I did this in the first place.  :-|

busjaeger commented on src/main/java/com/salesforce/dynamodbv2/mt/mappers/sharedtable/impl/RandomPartitioningConditionMapper.java:
```diff
@@ -175,15 +312,15 @@ static String getNextFieldPlaceholder(Map<String, String> expressionAttributeNam
-        final Pattern p = Pattern.compile(keyFieldName + " = ([^ ,]*)");
-        final Matcher m = p.matcher(conditionExpression);
-        return m.find() ? Optional.of(m.group(1)) : Optional.empty();
+        final Pattern p = Pattern.compile(fieldPlaceholder + " *" + COMPARISON_OP_PATTERN + " *(?<val>:[a-zA-Z0-9]+)");
```
This is probably pretty expensive. Switching to a grammar should help here.

hhkwong commented:
I had ZOS's subrepo use this branch and ran all the tests (if there's a way to skip mt-dynamo tests and run just ZOS tests that'd be great), and there were 3 new tests failing, because in applyForUpdate() I was assuming there'd always be an update expression. But actually it's okay to not have an update expression, and if the key is on an existing record, it's just a no-op. Also, while looking at the ZOS tests that failed I also realized that one can call UpdateItem on a non-existent key, and that should result in a new record being created -- I think I used to know this at some point when I added support for PATCH, because it turns out the dynamo behavior is like the OData PATCH behavior, but forgot about this. So, I fixed the condition mapper so it's okay to not have any update expression, and added relevant test cases in UpdateTest so we'd catch this in mt-dynamo in the future.

sbabu-salesforce commented:
hey @hhkwong, just a quick comment re: ```"if there's a way to skip mt-dynamo tests and run just ZOS tests that'd be great"```, see the `make test` target in the Makefile of zos. There's a profile you can pass to `mvn test` which will skip mt-dynamo's tests, and should only run ZOS tests
